fdbserver/CommitProxyServer.actor.cpp

```diff
@@ -2469,6 +2485,91 @@ ACTOR Future<Void> reportTxnTagCommitCost(UID myID,
 	}
 }
 
+namespace {
+struct ExpireServerEntry {
+	int64_t timeReceived;
+	int expectedCount = 0;
+	int receivedCount = 0;
+	bool initialized = false;
+};
+
+struct IdempotencyKey {
+	Version version;
+	uint8_t highOrderBatchIndex;
+	bool operator==(const IdempotencyKey& other) const {
+		return version == other.version && highOrderBatchIndex == other.highOrderBatchIndex;
+	}
+};
+
+} // namespace
+
+namespace std {
+template <>
+struct hash<IdempotencyKey> {
+	std::size_t operator()(const IdempotencyKey& key) const {
+		std::size_t seed = 0;
+		boost::hash_combine(seed, std::hash<Version>{}(key.version));
+		boost::hash_combine(seed, std::hash<uint8_t>{}(key.highOrderBatchIndex));
+		return seed;
+	}
+};
+
+} // namespace std
+
+ACTOR static Future<Void> idempotencyIdsExpireServer(
+    Database db,
+    PublicRequestStream<ExpireIdempotencyIdRequest> expireIdempotencyId,
+    FutureStream<ExpireIdempotencyKeyValuePairRequest> expireIdempotencyKeyValuePair,
+    Standalone<VectorRef<MutationRef>>* idempotencyClears) {
+	state std::unordered_map<IdempotencyKey, ExpireServerEntry> idStatus;
+	state std::unordered_map<IdempotencyKey, ExpireServerEntry>::iterator iter;
+	state int64_t purgeBefore;
+	state IdempotencyKey key;
+	state ExpireServerEntry* status = nullptr;
+	state Future<Void> purgeOld = Void();
+	loop {
+		choose {
+			when(ExpireIdempotencyIdRequest req = waitNext(expireIdempotencyId.getFuture())) {
+				key = IdempotencyKey{ req.commitVersion, req.batchIndexHighByte };
+				status = &idStatus[key];
+				status->receivedCount += 1;
```

Comment:
Is it guaranteed that received count is less than or equal to expected count? I could see that not being true if it were possible to get the "received" request before the "expected" request, but otherwise if that happens it would mean that we would have leaked the data. Since that shouldn't happen, we could possibly add an assert to that effect.

Reply:
As implemented it is guaranteed that received count is less than or equal to expected count, but that's only because we happen to still be in the same callstack/Task that sent the CommitID replies to clients. I think this should still work even if received is incremented before expected is set, so instead of the assert maybe I'll add some buggifies + code probes to make sure we handle this correctly?

Reply:
You could still possibly add an assert for this -- something like `ASSERT(expected == 0 || expected >= received)`

Reply:
Done